pkg/server/image_pull.go

```diff
@@ -17,14 +17,226 @@ limitations under the License.
 package server
 
 import (
-	"errors"
+	"encoding/json"
+	"fmt"
+	"time"
 
+	"github.com/containerd/containerd/content"
+	containerd_images "github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/remotes"
+	"github.com/containerd/containerd/remotes/docker"
+	"github.com/kubernetes-incubator/cri-containerd/pkg/metadata"
+	imagespec "github.com/opencontainers/image-spec/specs-go/v1"
 	"golang.org/x/net/context"
-
+	"golang.org/x/sync/errgroup"
 	"k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime"
 )
 
 // PullImage pulls an image with authentication config.
+// TODO(mikebrow): add authentication
+// TODO(mikebrow): harden api (including figuring out at what layer we should be blocking on duplicate requests.)
 func (c *criContainerdService) PullImage(ctx context.Context, r *runtime.PullImageRequest) (*runtime.PullImageResponse, error) {
-	return nil, errors.New("not implemented")
+	var (
+		err  error
+		size int64
+		desc imagespec.Descriptor
+	)
+
+	image := r.GetImage().Image
+
+	if desc, size, err = c.pullImage(ctx, image); err != nil {
+		return nil, fmt.Errorf("failed to pull image %q: %v", image, err)
+	}
+	digest := desc.Digest.String() // TODO(mikebrow): add truncIndex for image id
+
+	// TODO(mikebrow): pass a metadata struct to pullimage and fill in the tags/digests
+	// store the image metadata
+	var meta *metadata.ImageMetadata
+	meta, err = c.imageMetadataStore.Get(digest)
```

Comment:
Just `Create` directly.

Even you do a `Get` first, it could not avoid the creation failure, because there may be another thread creates the metadata after your `Get`.

Just `Create` direclty here, in future PR, we could let `Create` return `already exists` error, and ignore that error.